### PR TITLE
Handle claims conversion

### LINKED ISSUE #43: Use DateTime for date claims

### LINKED ISSUE #144: Support non-integer timestamps
RFC 7519 clearly states (page 5) that for NumericDate properties, "non-integer values can be represented", but your version 3.2.1 code appears to cast all timestamps to (int)!  For instance, Builder's setExpiration() and setNotBefore(); and ValidationData's setCurrentTime().  A whole second is an eternity in CPU time.  Can we please have decimal seconds?

### PR DESCRIPTION
**Pull request by lcobucci**
There're some weird stuff to improve @Ocramius (and I couldn't resist adding the last commit 😂).

Fixes #144 
Fixes #43

### CODE REVIEW

File: src/Token/Builder.php [hunk 1]
```diff
@@ -169,4 +170,29 @@ public function getToken(Signer $signer, Key $key): Plain
             new Signature($signature, $encodedSignature)
         );
     }
+
+    private function formatClaims(array $claims): array
+    {
+        if (isset($claims[RegisteredClaims::AUDIENCE]) && count($claims[RegisteredClaims::AUDIENCE]) === 1) {
```
Comment (Ocramius) [inline src/Token/Builder.php]: Please don't sneak in this optimisation here :-P

Comment (Ocramius) [inline src/Token/Builder.php]: Also, if you assume that it always is a packed array, simply use an `isset()` on index `0` and another one on index `1`

File: src/Token/Builder.php [hunk 2]
```diff
@@ -169,4 +170,29 @@ public function getToken(Signer $signer, Key $key): Plain
...
+                continue;
+            }
+
+            $claims[$claim] = $this->convertDate($claims[$claim]);
+        }
+
+        return $claims;
+    }
+
+    private function convertDate(DateTimeImmutable $date)
```
Comment (Ocramius) [inline src/Token/Builder.php]: Return type declaration needed

File: src/Token/Builder.php [hunk 3]
```diff
@@ -169,4 +170,29 @@ public function getToken(Signer $signer, Key $key): Plain
...
+    {
+        $seconds = $date->format('U');
+        $microseconds = $date->format('u');
+
+        return $microseconds === '000000' ? (int) $seconds : $seconds . '.' . $microseconds;
```
Comment (Ocramius) [inline src/Token/Builder.php]: Big assumption on the number of decimal digits. Instead, use `0 === (int) $microseconds`

File: src/Token/Builder.php (hunk 2 above)
Comment (Ocramius) [inline src/Token/Builder.php]: I see it's `int|string`. Can't it be `float`?

Comment (Ocramius) [inline src/Token/Builder.php]: Note that `json_encode()` does some trickery around floats, making them integers when no part is after the comma

Comment (Ocramius) [inline src/Token/Builder.php]: Ref: https://3v4l.org/Uq1CL

File: src/Token/Parser.php [hunk 4]
```diff
@@ -90,7 +91,30 @@ private function parseHeader(string $data): array
      */
     private function parseClaims(string $data): array
     {
-        return (array) $this->decoder->jsonDecode($this->decoder->base64UrlDecode($data));
+        $claims = (array) $this->decoder->jsonDecode($this->decoder->base64UrlDecode($data));
+
+        if (isset($claims[RegisteredClaims::AUDIENCE])) {
```
Comment (Ocramius) [inline src/Token/Parser.php]: Note that casting `null` to `(array)` yields an empty array. I don't know if that's valid for `$claims`, but worth considering

File: src/Token/Parser.php [hunk 5]
```diff
@@ -90,7 +91,30 @@ private function parseHeader(string $data): array
...
+            $claims[RegisteredClaims::AUDIENCE] = (array) $claims[RegisteredClaims::AUDIENCE];
+        }
+
+        foreach (RegisteredClaims::DATE_CLAIMS as $claim) {
+            if (!isset($claims[$claim])) {
```
Comment (Ocramius) [inline src/Token/Parser.php]: `areay_intersect_key()`?

File: src/Token/Parser.php [hunk 6]
```diff
@@ -90,7 +91,30 @@ private function parseHeader(string $data): array
...
+        return $claims;
+    }
+
+    private function convertDate(string $value): DateTimeImmutable
+    {
+        if (strpos($value, '.') === false) {
+            $value .= '.0';
+        }
+
+        return DateTimeImmutable::createFromFormat('U.u', $value);
```
Comment (Ocramius) [inline src/Token/Parser.php]: I'd use two different expressions for when a microtime is set or not: this way you avoid overwriting `$value`

File: src/Token/RegisteredClaims.php [hunk 7]
```diff
@@ -27,6 +27,8 @@
         self::SUBJECT
     ];
 
+    const DATE_CLAIMS = [self::ISSUED_AT, self::NOT_BEFORE, self::EXPIRATION_TIME];
```
Comment (Ocramius) [inline src/Token/RegisteredClaims.php]: One per line, please

File: src/Token/Builder.php (hunk 3 above)
Comment (lcobucci) [inline src/Token/Builder.php]: `DateTimeInterface#format('u')` always returns a 6 chars string. It can be compared to 0 though, there's no much difference

File: src/Token/Parser.php (hunk 4 above)
Comment (lcobucci) [inline src/Token/Parser.php]: I don't want to set if it doesn't exist in the original token

File: src/Token/Builder.php (hunk 2 above)
Comment (lcobucci) [inline src/Token/Builder.php]: I was with the old behaviour in my mind: https://3v4l.org/1Q7hh
Then when we convert it back to a `DateTime` we would have `DateTimeImmutable::createFromFormat('U.u', '1487372450.0231')` instead of `DateTimeImmutable::createFromFormat('U.u', '1487372450.023126')`, what can cause some weird errors.

File: src/Token/Builder.php (hunk 1 above)
Comment (lcobucci) [inline src/Token/Builder.php]: Yeah it's definitely better then `count()` in that case

File: src/Token/Builder.php (hunk 2 above)
Comment (Ocramius) [inline src/Token/Builder.php]: Ah, I see what you mean. If you can, you should cover the edge case you just mentioned in the test suite

File: src/Token/Builder.php (hunk 3 above)
Comment (Ocramius) [inline src/Token/Builder.php]: Yeah, but nothing prevents the engine from making this 7 chars, or similar, in future versions. This code is very fragile

File: src/Token/Parser.php (hunk 4 above)
Comment (Ocramius) [inline src/Token/Parser.php]: Makes sense

File: src/Token/Builder.php (hunk 3 above)
Comment (lcobucci) [inline src/Token/Builder.php]: Indeed

File: src/Token/Parser.php (hunk 5 above)
Comment (lcobucci) [inline src/Token/Parser.php]: In order to use it I would also have to do an `array_flip(RegisteredClaims::DATE_CLAIMS)` (or an `array_keys($claims)` and use `array_intersect()` instead) not sure if will help us...

File: src/Token/Builder.php (hunk 2 above)
Comment (lcobucci) [inline src/Token/Builder.php]: It's being covered on https://github.com/lcobucci/jwt/pull/171/files#diff-89e66398726c97dea8259f9c0d31b5f5R417

Comment (lcobucci) [inline src/Token/Builder.php]: But can be more explicit, ofc

File: src/Token/Parser.php (hunk 5 above)
Comment (Ocramius) [inline src/Token/Parser.php]: Or define the constants in the keys ;-)

**Comment by lcobucci** [comment]
@Ocramius could you please review it again?
